### PR TITLE
[SPARK-18325][SparkR][ML] SparkR ML wrappers example code and user guide

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?
* Add all R examples for ML wrappers which were added during 2.1 release cycle.
* Split the whole ```ml.R``` example file into individual example for each algorithm, which will be convenient for users to rerun them.
* Add corresponding examples to ML user guide.
* Update ML section of SparkR user guide. 

Note: MLlib Scala/Java/Python examples will be consistent, however, SparkR examples may different from them, since R users may use the algorithms in a different way, for example, using R ```formula``` to specify ```featuresCol``` and ```labelCol```.

## How was this patch tested?
Run all examples manually.

### CODE REVIEW

File: docs/ml-classification-regression.md [hunk 1]
```diff
@@ -75,6 +75,13 @@ More details on parameters can be found in the [Python API documentation](api/py
 {% include_example python/ml/logistic_regression_with_elastic_net.py %}
 </div>
 
+<div data-lang="r" markdown="1">
+
+More details on parameters can be found in the [R API documentation](api/R/spark.logit.html).
```
Comment (wangmiao1981) [inline docs/ml-classification-regression.md]: Change to "Refer to the [R API docs]... for more details"? For consistence.

File: docs/ml-classification-regression.md [hunk 2]
```diff
@@ -847,6 +893,14 @@ Refer to the [`IsotonicRegression` Python docs](api/python/pyspark.ml.html#pyspa
 
 {% include_example python/ml/isotonic_regression_example.py %}
 </div>
+
+<div data-lang="r" markdown="1">
+
+Refer to the [`IsotonicRegression` R API docs](api/R/spark.isoreg.html) for more details on the API.
```
Comment (wangmiao1981) [inline docs/ml-classification-regression.md]: Delete "on the API" for consistence?

File: docs/sparkr.md [hunk 3]
```diff
@@ -512,39 +512,33 @@ head(teenagers)
...
-### Naive Bayes Model
-
-[spark.naiveBayes()](api/R/spark.naiveBayes.html) fits a Bernoulli naive Bayes model against a SparkDataFrame. Only categorical data is supported.
-{% include_example naiveBayes r/ml.R %}
-
-### KMeans Model
+SparkR supports the following machine learning algorithms currently:
+
+* `spark.glm` or `glm`: `Generalized Linear Model`
+* `spark.survreg`: `Accelerated Failure Time (AFT) Survival Regressio Model`
```
Comment (wangmiao1981) [inline docs/sparkr.md]: Typo: `Regressio` -> `Regression`

File: examples/src/main/r/ml/isoreg.R [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+# Fit an isotonic regression model with spark.isoreg
+model <- spark.isoreg(training, label ~ features, isotonic = FALSE)
+
+# Model summary
+summary(model)
+
+# Prediction
+predictions <- predict(model, test)
+showDF(predictions)
+# $example off$
```
Comment (wangmiao1981) [inline examples/src/main/r/ml/isoreg.R]: newline

File: examples/src/main/r/ml/ml.R [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+train <- function(family) {
+  model <- glm(Sepal.Length ~ Sepal.Width + Species, iris, family = family)
+  summary(model)
+}
+model.summaries <- spark.lapply(families, train)
+
+# Print the summary of each model
+print(model.summaries)
+
+
```
Comment (wangmiao1981) [inline examples/src/main/r/ml/ml.R]: Extra blank line

**Review by wangmiao1981 — COMMENTED** [review]
Minor comments on format and typo etc.

File: docs/sparkr.md [hunk 6]
```diff
@@ -512,39 +512,33 @@ head(teenagers)
...
-Under the hood, SparkR uses MLlib to train the model.
-Users can call `summary` to print a summary of the fitted model, [predict](api/R/predict.html) to make predictions on new data, and [write.ml](api/R/write.ml.html)/[read.ml](api/R/read.ml.html) to save/load fitted models.
-SparkR supports a subset of the available R formula operators for model fitting, including ‘~’, ‘.’, ‘:’, ‘+’, and ‘-‘.
-
 ## Algorithms
 
-### Generalized Linear Model
-
-[spark.glm()](api/R/spark.glm.html) or [glm()](api/R/glm.html) fits generalized linear model against a Spark DataFrame.
-Currently "gaussian", "binomial", "poisson" and "gamma" families are supported.
```
Comment (felixcheung) [inline docs/sparkr.md]: looks like we would be missing out some R specific things from this delete?

File: docs/sparkr.md [hunk 7]
```diff
@@ -512,39 +512,33 @@ head(teenagers)
...
-{% include_example glm r/ml.R %}
-
-### Accelerated Failure Time (AFT) Survival Regression Model
-
-[spark.survreg()](api/R/spark.survreg.html) fits an accelerated failure time (AFT) survival regression model on a SparkDataFrame.
-Note that the formula of [spark.survreg()](api/R/spark.survreg.html) does not support operator '.' currently.
```
Comment (felixcheung) [inline docs/sparkr.md]: another R specific info that would be deleted?

File: examples/src/main/r/ml/randomForest.R [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+# Random forest classification model
+
+# $example on:classification$
+# Load training data
+df <- read.df("data/mllib/sample_libsvm_data.txt", source = "libsvm")
+training <- df
+test <- df
+
+# Fit a random forest classification model with spark.randomForest
+model <- spark.randomForest(training, label ~ features, "classification", numTrees=10)
```
Comment (felixcheung) [inline examples/src/main/r/ml/randomForest.R]: nit: I would put space around, ie. `numTrees = 10` instead

Comment (felixcheung) [inline examples/src/main/r/ml/randomForest.R]: ditto below

File: examples/src/main/r/ml/lda.R [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+sparkR.session(appName = "SparkR-ML-lda-example")
+
+# $example on$
+# Load training data
+df <- read.df("data/mllib/sample_lda_libsvm_data.txt", source = "libsvm")
+training <- df
+test <- df
+
+# Fit a latent dirichlet allocation model with spark.lda
+model <- spark.lda(training, k=10, maxIter=10)
```
Comment (felixcheung) [inline examples/src/main/r/ml/lda.R]: nit: please put space, ie. `k = 10, maxIter = 10`

**Comment by felixcheung** [comment]
this is great, thanks! btw, how are these examples getting run? is there a way to know if the examples are broken because of API changes?

**Review by yanboliang — COMMENTED** [review]
@wangmiao1981 @felixcheung I addressed your comments. Thanks for your reviewing.

**Comment by yanboliang** [comment]
@felixcheung That's a good question. We don't run all MLlib examples for testing until [SPARK-12347](https://issues.apache.org/jira/browse/SPARK-12347) was resolved.

**Comment by felixcheung** [comment]
hmm, yea, I suspect these example can be easily forgotten while changes are made - they could end up getting out of sync easily

**Comment by yanboliang** [comment]
Yeah, I have the same concerns, so we should push forward SPARK-12347 and it should not be very hard to do that.

File: docs/sparkr.md [hunk 10]
```diff
@@ -512,39 +512,33 @@ head(teenagers)
...
-[spark.naiveBayes()](api/R/spark.naiveBayes.html) fits a Bernoulli naive Bayes model against a SparkDataFrame. Only categorical data is supported.
-{% include_example naiveBayes r/ml.R %}
-
-### KMeans Model
+SparkR supports the following machine learning algorithms currently:
+
+* [`spark.glm`](api/R/spark.glm.html) or [`glm`](api/R/glm.html): [`Generalized Linear Model`](ml-classification-regression.html#generalized-linear-regression)
+* [`spark.survreg`](api/R/spark.survreg.html): [`Accelerated Failure Time (AFT) Survival Regression Model`](ml-classification-regression.html#survival-regression)
+* [`spark.naiveBayes`](api/R/spark.naiveBayes.html): [`Naive Bayes Model`](ml-classification-regression.html#naive-bayes)
+* [`spark.kmeans`](api/R/spark.kmeans.html): [`KMeans Model`](ml-clustering.html#k-means)
```
Comment (felixcheung) [inline docs/sparkr.md]: nit: `k-Means`? I think we have the `-` in general

File: examples/src/main/r/ml/ml.R [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+unlink(modelPath)
+# $example off:read_write$
+############################ fit models with spark.lapply #####################################
+
+# Perform distributed training of multiple models with spark.lapply
+families <- c("gaussian", "poisson")
+train <- function(family) {
+  model <- glm(Sepal.Length ~ Sepal.Width + Species, iris, family = family)
+  summary(model)
+}
```
Comment (felixcheung) [inline examples/src/main/r/ml/ml.R]: might want to use this for example instead
https://github.com/apache/spark/blame/master/R/pkg/vignettes/sparkr-vignettes.Rmd#L394

File: examples/src/main/r/ml/glm.R [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+# Prediction
+gaussianPredictions <- predict(gaussianGLM, gaussianTestDF)
+showDF(gaussianPredictions)
+
+# Fit a generalized linear model with glm (R-compliant)
+gaussianGLM2 <- glm(Sepal_Length ~ Sepal_Width + Species, gaussianDF, family = "gaussian")
+summary(gaussianGLM2)
+
+# Fit a generalized linear model of family "binomial" with spark.glm
+binomialDF <- filter(irisDF, irisDF$Species != "setosa")
```
Comment (felixcheung) [inline examples/src/main/r/ml/glm.R]: perhaps to add a comment on why we are filtering out `setosa`?

**Comment by felixcheung** [comment]
A couple more comments, LGTM otherwise. Thanks

File: docs/sparkr.md (hunk 10 above)
Comment (yanboliang) [inline docs/sparkr.md]: Done.

File: examples/src/main/r/ml/ml.R (hunk 11 above)
Comment (felixcheung) [inline examples/src/main/r/ml/ml.R]: I mean running 2 families on glm in parallel is not a very useful or practical example. The link [above](https://github.com/apache/spark/blame/master/R/pkg/vignettes/sparkr-vignettes.Rmd#L394), running a (potentially longer) list of costs might be a better example to include here.

Comment (yanboliang) [inline examples/src/main/r/ml/ml.R]: Good suggestion, I'll update it. Thanks.

**Comment by felixcheung** [comment]
LGTM. Looks like we are locked down for 2.1. Good to have with all the new examples but seems like a lot of code (example) changes?

**Comment by yanboliang** [comment]
Yeah, 2.1 RC2 has been cut, but I think we still need to merge it into branch-2.1, since it's possible we have new RC. Even if RC2 passed voting, we will have 2.1.1, 2.1.2, ... which need this examples and docs. Merged into master and branch-2.1. Thanks for all reviewing.

**Comment by jkbradley** [comment]
@felixcheung This is fine to merge since it is for docs/examples only.  But in general, we should be better about adding these right away after adding the R APIs, rather than doing them all at the end of the release cycle.  If a contributor adds a new API, they should follow up immediately with the docs.

**Comment by felixcheung** [comment]
Agreed. We will definitely keep that in mind but there is certainly a lot going in working the API already.

**Pull request closed without merging**